feat[parser]: support underscores in numeric literals

z80dev opened the pull request:
### What I did

add support for underscores as visual separators in numeric literals
(PEP 515), allowing e.g. `100_000` in place of `100000`.

### How I did it

update the Lark grammar to accept underscores in decimal, hex, octal,
binary, and float literals. break the float regex into composable
sub-parsers (`_DIGITPART`, `_FRACTION`, `_EXPONENT`) for
readability per PEP 515. also accept uppercase prefixes (`0X`, `0O`,
`0B`) per the Python literal grammar.

in the AST annotator, hoist underscore stripping to the top of
`visit_Num` so all numeric branches benefit from a single call.

### How to verify it

`pytest tests/functional/codegen/types/numbers/test_numeric_literals_underscores.py`

### Commit message

```
feat[parser]: support underscores in numeric literals

add support for underscores as visual separators in all numeric literal
forms (decimal, hex, octal, binary, float), following PEP 515. this
allows writing e.g. `1_000_000` or `0xFF_FF` for readability.

update the Lark grammar to accept underscores via `(_?digit)+` groups.
break the float number regex into composable sub-parsers (_DIGITPART,
_FRACTION, _EXPONENT) for readability. accept uppercase prefixes (0X,
0O, 0B) per the Python literal grammar.

in the AST annotator, hoist underscore stripping to the top of
visit_Num so it applies uniformly to all numeric branches.
```

### Description for the changelog

support underscores in numeric literals (e.g. `1_000_000`, `0xFF_FF`)

### Cute Animal Picture

![cute animal](https://i.insider.com/60d0aae33093db0019702380?width=700)

z80dev committed:
support '_' separator in dec/oct/float numbers

Allows 100_000 in place of 100000

z80dev committed:
feat: add tests confirming underscore sep

z80dev committed:
fix: simplify unit tests

z80dev committed:
Merge remote-tracking branch 'origin/master' into lark-allow-underscore-numerical-sep

charles-cooper reviewed (changes requested):
looks like the new grammar breaks some tests

charles-cooper committed:
Merge branch 'master' into lark-allow-underscore-numerical-sep

charles-cooper committed:
Merge branch 'master' into lark-allow-underscore-numerical-sep

charles-cooper committed:
fix grammar rules

charles-cooper committed:
update vyper parser and tests

charles-cooper committed:
Merge branch 'master' into lark-allow-underscore-numerical-sep

charles-cooper committed:
Merge branch 'master' into lark-allow-underscore-numerical-sep

cyberthirst commented:
Please add tests for all supported types. Also add tests for edgecases like _0, 10__0, 10_,  0xaa__aa etc

charles-cooper committed:
add some test cases

charles-cooper committed:
move to codegen tests

Sporarum commented on vyper/ast/grammar.lark:
```diff
@@ -309,11 +310,11 @@ _NEWLINE: ( /\r?\n[\t ]*/ | COMMENT )+
+DEC_NUMBER: /0|[1-9](_?[0-9])*/
+HEX_NUMBER.2: /0x_?[\da-f](_?[\da-f])*/i
+OCT_NUMBER.2: /0o_?[0-7](_?[0-7])*/
+BIN_NUMBER.2 : /0b_?[0-1](_?[0-1])*/i
```
Isn't this equivalent ?
```suggestion
HEX_NUMBER.2: /0x(_?[\da-f])+/i
OCT_NUMBER.2: /0o(_?[0-7])+/
BIN_NUMBER.2 : /0b(_?[0-1])+/i
```

Furthermore, we should also accept uppercase:
```suggestion
HEX_NUMBER.2: /0[xX](_?[\da-f])+/i
OCT_NUMBER.2: /0[oO](_?[0-7])+/
BIN_NUMBER.2 : /0[bB](_?[0-1])+/i
```

See https://peps.python.org/pep-0515/#literal-grammar

Sporarum reviewed (commented):
Very nice, a couple comments but nothing major

Don't forget to add tests for the uppercase variants as well

cyberthirst commented on vyper/ast/parse.py:
```diff
@@ -490,6 +490,8 @@ def visit_Num(self, node):
 
```
agreed, i think it's better to treat all values consistently and have them share the property of having the underscores replaced

charles-cooper committed:
address review: simplify grammar, hoist underscore stripping

- grammar.lark: simplify HEX/OCT/BIN regexes, add uppercase prefix support (0X, 0O, 0B)
- grammar.lark: break FLOAT_NUMBER into _DIGITPART/_FRACTION/_EXPONENT sub-parsers (PEP 515)
- parse.py: hoist underscore stripping to top of visit_Num instead of per-branch

charles-cooper committed:
Merge branch 'master' into lark-allow-underscore-numerical-sep
